File: backend/src/directory/directory.service.ts
import { Injectable, OnModuleInit } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import * as fs from 'fs';
import * as path from 'path';

export const defaultTeams = [
  'TEAM ANCHOR',
  'TEAM ALPHA',
  'TITAN LITIGATION',
  'MARITIME PRACTICE GROUP',
  'CORPORATE DISPUTE TEAM',
];

export const defaultCourts = [
  'HIGH COURT BENIN CITY',
  'HIGH COURT OKADA',
  'HIGH COURT EKIADOLOR',
  'EKIADOLOR MAGISTRATE COURT',
  'FEDERAL HIGH COURT',
  'HIGH COURT',
  'MAGISTRATE COURT OGBESON',
  'MAGISTRATE COURT OREDO',
  'MAGISTRATE COURT EGOR',
  'HIGH COURT WARRI',
  'HIGH COURT ABUDU',
  'FEDERAL HIGH COURT BENIN',
  'HIGH COURT BENIN',
  'APPEAL COURT BENIN CITY',
  'NATIONAL INDUSTRIAL COURT BENIN CITY',
  'AREA CUSTOMARY COURT EHOR',
  'CUSTOMARY COURT URHONIGBE',
  'HIGH COURT EHOR',
];

@Injectable()
export class DirectoryService implements OnModuleInit {
  private filePath = path.join(process.cwd(), 'data', 'directory.json');
  private courts: string[] = [];
  private teams: string[] = [];
  private isLoaded = false;

  constructor(private readonly prisma: PrismaService) {}

  async onModuleInit() {
    await this.loadData();
  }

  private async loadData() {
    // 1. Prioritize reading from Database (Firestore / persistent store)
    try {
      const record = await this.prisma.directory.findUnique({
        where: { id: 'settings' },
      });
      if (record && (Array.isArray(record.teams) || Array.isArray(record.courts))) {
        this.teams =
          Array.isArray(record.teams) && record.teams.length > 0
            ? record.teams
            : [...defaultTeams];
        this.courts =
          Array.isArray(record.courts) && record.courts.length > 0
            ? record.courts
            : [...defaultCourts];
        this.isLoaded = true;
        this.saveToFile();
        return;
      }
    } catch (dbErr: any) {
      console.warn(
        '[DirectoryService] Could not read directory from DB, checking local file:',
        dbErr?.message || dbErr,
      );
    }

    // 2. Fallback to local JSON file
    try {
      if (fs.existsSync(this.filePath)) {
        const raw = fs.readFileSync(this.filePath, 'utf-8');
        const parsed = JSON.parse(raw);
        this.courts =
          Array.isArray(parsed.courts) && parsed.courts.length > 0
            ? parsed.courts
            : [...defaultCourts];
        this.teams =
          Array.isArray(parsed.teams) && parsed.teams.length > 0
            ? parsed.teams
            : [...defaultTeams];
        this.isLoaded = true;
        await this.syncToDb();
        return;
      }
    } catch (fileErr: any) {
      console.warn(
        '[DirectoryService] Could not read directory from file:',
        fileErr?.message || fileErr,
      );
    }

    // 3. Fallback to defaults
    this.courts = [...defaultCourts];
    this.teams = [...defaultTeams];
    this.isLoaded = true;
    await this.persistData();
  }

  private async syncToDb() {
    try {
      const existing = await this.prisma.directory.findUnique({
        where: { id: 'settings' },
      });
      if (existing) {
        await this.prisma.directory.update({
          where: { id: 'settings' },
          data: {
            teams: this.teams,
            courts: this.courts,
          },
        });
      } else {
        await this.prisma.directory.create({
          data: {
            id: 'settings',
            teams: this.teams,
            courts: this.courts,
          },
        });
      }
    } catch (dbErr: any) {
      console.warn(
        '[DirectoryService] DB sync warning:',
        dbErr?.message || dbErr,
      );
    }
  }

  private async persistData() {
    this.saveToFile();
    await this.syncToDb();
  }

  private saveToFile() {
    try {
      const dir = path.dirname(this.filePath);
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
      fs.writeFileSync(
        this.filePath,
        JSON.stringify({ courts: this.courts, teams: this.teams }, null, 2),
        'utf-8',
      );
    } catch (e) {
      console.error('[DirectoryService] Failed to save directory data to file:', e);
    }
  }

  async getCourts(): Promise<string[]> {
    if (!this.isLoaded) await this.loadData();
    return [...this.courts];
  }

  async addCourt(name: string): Promise<string[]> {
    if (!this.isLoaded) await this.loadData();
    const clean = name.trim().toUpperCase();
    if (clean && !this.courts.includes(clean)) {
      this.courts.push(clean);
      await this.persistData();
    }
    return this.getCourts();
  }

  async removeCourt(name: string): Promise<string[]> {
    if (!this.isLoaded) await this.loadData();
    const clean = name.trim().toUpperCase();
    this.courts = this.courts.filter((c) => c.toUpperCase() !== clean);
    await this.persistData();
    return this.getCourts();
  }

  async getTeams(): Promise<string[]> {
    if (!this.isLoaded) await this.loadData();
    return [...this.teams];
  }

  async addTeam(name: string): Promise<string[]> {
    if (!this.isLoaded) await this.loadData();
    const clean = name.trim().toUpperCase();
    if (clean && !this.teams.includes(clean)) {
      this.teams.push(clean);
      await this.persistData();
    }
    return this.getTeams();
  }

  async removeTeam(name: string): Promise<string[]> {
    if (!this.isLoaded) await this.loadData();
    const clean = name.trim().toUpperCase();
    this.teams = this.teams.filter((t) => t.toUpperCase() !== clean);
    await this.persistData();
    return this.getTeams();
  }
}

File: mobile/lib/services/directory_service.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import 'api_service.dart';
import '../config/api_config.dart';

class DirectoryService {
  static const String _teamsKey = 'midlex_litigation_teams';
  static const String _courtsKey = 'midlex_courts_directory';
  static const String _assignmentsKey = 'midlex_case_assignments';

  static const List<String> defaultTeams = [
    'TEAM ANCHOR',
    'TEAM ALPHA',
    'TITAN LITIGATION',
    'MARITIME PRACTICE GROUP',
    'CORPORATE DISPUTE TEAM',
  ];

  static const List<String> defaultCourts = [
    'HIGH COURT BENIN CITY',
    'HIGH COURT OKADA',
    'HIGH COURT EKIADOLOR',
    'EKIADOLOR MAGISTRATE COURT',
    'FEDERAL HIGH COURT',
    'HIGH COURT',
    'MAGISTRATE COURT OGBESON',
    'MAGISTRATE COURT OREDO',
    'MAGISTRATE COURT EGOR',
    'HIGH COURT WARRI',
    'HIGH COURT ABUDU',
    'FEDERAL HIGH COURT BENIN',
    'HIGH COURT BENIN',
    'APPEAL COURT BENIN CITY',
    'NATIONAL INDUSTRIAL COURT BENIN CITY',
    'AREA CUSTOMARY COURT EHOR',
    'CUSTOMARY COURT URHONIGBE',
    'HIGH COURT EHOR',
  ];

  // Teams Management (Synced with Backend DB)
  static Future<List<String>> getTeams() async {
    final prefs = await SharedPreferences.getInstance();
    try {
      final res = await ApiService.get(ApiConfig.directoryTeams);
      if (res is List && res.isNotEmpty) {
        final teams = res.map((e) => e.toString().trim().toUpperCase()).toSet().toList();
        await prefs.setStringList(_teamsKey, teams);
        return teams;
      }
    } catch (_) {}

    final list = prefs.getStringList(_teamsKey);
    if (list == null || list.isEmpty) {
      await prefs.setStringList(_teamsKey, defaultTeams);
      return List.from(defaultTeams);
    }
    return list;
  }

  static Future<List<String>> addTeam(String teamName) async {
    final clean = teamName.trim().toUpperCase();
    if (clean.isEmpty) return await getTeams();
    final prefs = await SharedPreferences.getInstance();

    try {
      final res = await ApiService.post(ApiConfig.directoryTeams, {'name': clean});
      if (res is List && res.isNotEmpty) {
        final teams = res.map((e) => e.toString().trim().toUpperCase()).toSet().toList();
        await prefs.setStringList(_teamsKey, teams);
        return teams;
      }
    } catch (_) {}

    final teams = await getTeams();
    if (!teams.contains(clean)) {
      teams.add(clean);
      await prefs.setStringList(_teamsKey, teams);
    }
    return teams;
  }

  static Future<List<String>> removeTeam(String teamName) async {
    final clean = teamName.trim().toUpperCase();
    final prefs = await SharedPreferences.getInstance();

    try {
      final res = await ApiService.delete('${ApiConfig.directoryTeams}/${Uri.encodeComponent(clean)}');
      if (res is List && res.isNotEmpty) {
        final teams = res.map((e) => e.toString().trim().toUpperCase()).toSet().toList();
        await prefs.setStringList(_teamsKey, teams);
        return teams;
      }
    } catch (_) {}

    final teams = await getTeams();
    teams.removeWhere((t) => t.toUpperCase() == clean);
    await prefs.setStringList(_teamsKey, teams);
    return teams;
  }

  // Courts Management (Synced with Backend DB)
  static Future<List<String>> getCourts() async {
    final prefs = await SharedPreferences.getInstance();
    try {
      final res = await ApiService.get(ApiConfig.directoryCourts);
      if (res is List && res.isNotEmpty) {
        final courts = res.map((e) => e.toString().trim().toUpperCase()).toSet().toList();
        await prefs.setStringList(_courtsKey, courts);
        return courts;
      }
    } catch (_) {}

    final list = prefs.getStringList(_courtsKey);
    if (list == null || list.isEmpty) {
      await prefs.setStringList(_courtsKey, defaultCourts);
      return List.from(defaultCourts);
    }
    return list;
  }

  static Future<List<String>> addCourt(String courtName) async {
    final clean = courtName.trim().toUpperCase();
    if (clean.isEmpty) return await getCourts();
    final prefs = await SharedPreferences.getInstance();

    try {
      final res = await ApiService.post(ApiConfig.directoryCourts, {'name': clean});
      if (res is List && res.isNotEmpty) {
        final courts = res.map((e) => e.toString().trim().toUpperCase()).toSet().toList();
        await prefs.setStringList(_courtsKey, courts);
        return courts;
      }
    } catch (_) {}

    final courts = await getCourts();
    if (!courts.contains(clean)) {
      courts.add(clean);
      await prefs.setStringList(_courtsKey, courts);
    }
    return courts;
  }

  static Future<List<String>> removeCourt(String courtName) async {
    final clean = courtName.trim().toUpperCase();
    final prefs = await SharedPreferences.getInstance();

    try {
      final res = await ApiService.delete('${ApiConfig.directoryCourts}/${Uri.encodeComponent(clean)}');
      if (res is List && res.isNotEmpty) {
        final courts = res.map((e) => e.toString().trim().toUpperCase()).toSet().toList();
        await prefs.setStringList(_courtsKey, courts);
        return courts;
      }
    } catch (_) {}

    final courts = await getCourts();
    courts.removeWhere((c) => c.toUpperCase() == clean);
    await prefs.setStringList(_courtsKey, courts);
    return courts;
  }

  // Case Directory Custom Attributes (Suit No, Court, Litigation Team)
  static Future<Map<String, dynamic>> getCaseAssignment(String caseId) async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString('${_assignmentsKey}_$caseId');
    if (raw != null) {
      try {
        return jsonDecode(raw) as Map<String, dynamic>;
      } catch (e) {
        return {};
      }
    }
    return {};
  }

  static Future<void> saveCaseAssignment(
    String caseId, {
    String? suitNumber,
    String? court,
    String? litigationTeam,
  }) async {
    final prefs = await SharedPreferences.getInstance();
    final existing = await getCaseAssignment(caseId);
    if (suitNumber != null) existing['suitNumber'] = suitNumber;
    if (court != null) existing['court'] = court;
    if (litigationTeam != null) existing['litigationTeam'] = litigationTeam;
    await prefs.setString('${_assignmentsKey}_$caseId', jsonEncode(existing));
  }

  static Future<void> exportPdf(List<Map<String, dynamic>> items, {required String title}) async {
    // Export Directory items as PDF document
  }

  static Future<void> exportDocx(List<Map<String, dynamic>> items, {required String title}) async {
    // Export Directory items as Word DOCX document
  }
}
